### ISSUE TITLE
Confused about the objects that are returned and how to access them

### ISSUE THREAD

**Issue by dartheditous** [jlevers/selling-partner-api#506]
So, I'm not very up-to-date with objects and models and so forth. I get around this by json_encoding the objects I get back from calls and then json_decoding them to an array. Makes things nice and simple.

But I'm sure this isn't the "right" way to do it. The trouble is I find the whole thing kind of opaque and it isn't very clear what I'm supposed to do.

For example: getFeeds and getShipments. I can do this:

```
$result = $api->getFeeds(...);
print_r($result->feeds);
```

and it prints the feeds. Great. But compare that to getShipments:

```
$result = $api->getShipments(...)
print_r($result); // no problem
print_r($result->payload); // also no problem
print_r($result->payload->shipment_data; // this results in "<b>Fatal error</b>:  Uncaught Error: Call to undefined method SellingPartnerApi\Model\FbaInboundV0\GetShipmentsResult::getShipment_data() in /var/www/html/lib/sp-api/vendor/jlevers/selling-partner-api/lib/Model/BaseModel.php:267"
```

Instead I have to do:

```
print_r($result->payload->ShipmentData);
```

but it's not clear to me how I'm supposed to know this. I don't see any reference to ShipmentData (instead of shipment_data) in the documentation for the response/result objects.

Is this just basic object/method/property stuff that I haven't yet understood, or is it some kind of inconsistency in the API?

I just dislike the fact that json_decode(json_encode($result), true); makes thing so much easier...

If anyone understands why I'm confused and can point me in the direction of some further reading I would be grateful.

**Comment by misterakko** [comment]
Basically, Amazon SP-API returns a JSON and the library does its best to decode it and return a strongly typed, structured, multi-level object. You are tempted to do the reverse job which is, at best, a waste of CPU cycles.

On the other hand... The behaviour of the library is partly due to the fact that it uses a component which processes every JSON it gets the same way. So, for example, an API call which returns sixty different pieces of information will be corralled into an object with sixty getter functions and sixty setter functions, even if it makes zero sense to, say, individually set one field in an object which is always returned from the API and never an input for the API.

In most cases it's tolerable because the call returns something which is moderate in size and complexity.

**Comment by jlevers** [comment]
> Instead I have to do:
>
>  ```
>  print_r($result->payload->ShipmentData);
>  ```
> but it's not clear to me how I'm supposed to know this. I don't see any reference to ShipmentData (instead of shipment_data) in the documentation for the response/result objects.

TL;DR: I used the default OpenAPI [PHP templates](https://github.com/swagger-api/swagger-codegen/tree/master/modules/swagger-codegen/src/main/resources/php) to generate this library initially, and missed some of the problems with that strategy.

This inconsistency is my fault – when I made the first couple versions of this project, I hadn't used the OpenAPI generate  because of some knowledge I lacked about how the OpenAPI generate I'm using works, when I made the first couple versions of the project. (Basically, OpenAPI defaults to `snake_case` for all property names, even if the OpenAPI models use `camelCase`, and I thought that switching the casing after a few versions would cause more confusion than it would save.)

As @misterakko pointed out, the way the classes in this library work is a bit heavy-handed, and I've been playing around with the idea of making a more streamlined version...but that would result in massive breaking changes, and the benefit of the way things are set up now is that it makes it relatively easy to validate the structure of an object.
